README
Bio-Das::Lite
===========

Bio::Das::Lite is an implementation of the BioDas protocol, http://biodas.org/
for the retrieval of biological data from XML sources over HTTP.

Bio::Das::Lite is designed as a lightweight and more forgiving alternative to
the client/retrieval/parsing components of Bio::Das. Bio::Das::Lite itself is
not a drop-in replacement for Bio::Das but it can be subclassed to do so
(and probably will in a short-term future release).


NOTE
Results are returned hashed against the query URL which caused them.
This means that sometimes you receive more results than you might expect
(number of sources * number of segments)


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


DEPENDENCIES

This module requires these other modules and libraries:

LWP::Parallel::UserAgent
SOAP::Lite is required for non-DAS/2 registry support

COPYRIGHT AND LICENCE

Copyright (c) 2007 by Roger Pettett, Genome Research Ltd.
http://www.sanger.ac.uk/

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Bio::Das::Lite.

2.10 DAS 1.6 spec, all patches from Andy

2.05 Probably the last release on the DAS 1.53 spec

2.04 Patches to catch out-of-date versions of libcurl.
     Refactored some routines to reduce complexity.

2.01 Bugfix for authenticating proxies.
     No longer appends ? to DAS request URIs without query params.
     Supports links with no descriptions.

2.00 Switched to WWW::Curl::Simple (from LWP::Parallel::UserAgent)

1.61 Minor: Fixed prereq (for CPANTS prereq_matches_use)

1.60 Latest LWP has broken support for "parse_head"
     function. Das::Lite attempts to correct for this by squirting an
     ersatz version into the symbol table under LWP::Protocol, where
     it used to be. A particularly ugly hack!

1.58  Added 'no_proxy' configurable parameter.
      Added 'max_hosts' configurable parameter.
      Supports the X-DAS-Status header.
      Supports score-based stylesheets (see DAS 1.53E spec).
      Fixes for perlcritic.
      Fixed issues of failings tests caused by retired DAS servers.

1.53  kwalitee updates (Test::Pod::Coverage & LICENSE in Makefile.PL)

1.50  added 'sequence' patch for alignments, patch from jt6

1.49  new 'maxbins' parameter for feature requests:

      In recent months more and more das sources appeared that provide
      very condensed value based data. Sometimes it is a feature per
      every 50 basepairs or even a feature per amino acid. It causes
      two major problems:
      1 - amount of XML being transferred is huge
      2 - Ensembl can only do simple merge of the features, i.e just
      pick max or min value - and in some cases das sources providers
      want to use a completely different dataset for a different size
      and resolution of the viewed region.

      To solve these problems we have introduced a new `maxbins`
      parameter to the `features` das request.  Using this parameter
      Ensembl tells all das sources what is the width in pixels of the
      image used to display the features.  Then the das sources smart
      enough to understand the parameter will merge the features
      themselves and for the rest Ensembl will do the work.

1.48  'dna' request support. Minor refactorings and cleanups for perl-critic

1.47  Reworked tests, a couple of fixes for structure handling

1.44  Package namespace changed to Bio::Das::Lite
      -compat tests added for compatibility with deprecated DasLite.pm
      Codebase cleaned up to make Perl::Critic happy
      Protein structure support, patch from rdf
      'sources' request support, patch from rdf

1.41  Protein alignment support, patch from rdf

1.40  Cleanup for Perl::Critic purposes

1.39  Minor typographical tidying up, some more tests

1.38  Minor tweaks to http_proxy - support for http://user:pass@proxy.example.com:nnnn/ format http_proxy

1.33  Minor tweaks to proxy_* methods; Addition of configurable user_agent

1.32  Bugfixes: Cache cleaning for performing multiple queries using persistent objects
                Regex fixes for group notes and links
                Initial support for authenticating proxies (proxy_user, proxy_pass)

1.27  Switched over to CVS version to make things easier for me to track
      Minor revisions. Some more tests
 
0.15  Includes registry support.
      See 'new_from_registry', 'registry' and 'registry_sources' methods.

0.14  Fix for entry_points structure and a new test for it
      Added a new way to use the constructor: my $das = Bio::DasLite->new("http://server/das/source");
      Added a convenience side-effect - give segment data in responses for segments without features in

0.13  Absolutely minor change to include LWP::P::UA max_req connection throttling
      Upgrade strongly recommended

0.12  Reworked all the tests.
      Tweaked basename regex to allow more flexibility in
      specifying services rather than dsns, i.e.
      'dsn' => 'http://das.mysite.com/das'
      rather than
      'dsn' => 'http://das.mysite.com/das/dsn'

0.11  Added a little extra debug, some extra eval traps, improved url processing
      and a couple of new WTSI extensions to the spec

0.10  Fixed dsns DAS request (source basename)

0.09  Fixed broken 'types' request

0.08  Added multi-link, multi-note, multi-group-link and multi-group-note support
      Added X-Forwarded-For HTTP header for IP-based ACLs
      Added passthrough for HTTP header data

0.07  Internal development version

0.06  Fixes for the regex/stack problems in 0.05.
      Added callback support for stylesheet and features calls

0.05  Stylesheet support. Seems to tickle some perl regex/stack bugs
      on certain platforms (x86 Linux, Alpha Tru64). It works fine on my G4.

0.04  Internal version. Added duplication of segment data across
      features in the 'features' call.

0.03  Allowed for calls (namely 'features') with an arrayref of
      segments, i.e. $das->features(["1:1,1000", "2:1,1000"]);
      Minor modifications following some benchmarking and profiling.
      Results are now cached (per Bio::DasLite instance) against
      query URL by default.

0.02  Following feedback, data structures returned by:
      dsn, basename, dsns, entry_points, types, features, sequence
      have now been homogenised across the board. Results are now
      always hashed against the query which was made, regardless
      of how many sources are configured.
      Also added a few more tests including feature-by-id

0.01  Fri Aug 19 16:38:05 2005
	- original version; created by h2xs 1.23 with options
		-cfn Bio::DasLite



t/FileStub.pm
#########
# Author:        rmp
# Last Modified: $Date: 2010-03-24 19:29:46 +0000 (Wed, 24 Mar 2010) $ $Author: zerojinx $
# Id:            $Id: FileStub.pm 19 2010-03-24 19:29:46Z zerojinx $
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/t/FileStub.pm,v $
# $HeadURL: https://bio-das-lite.svn.sourceforge.net/svnroot/bio-das-lite/tags/spec-1.53/t/FileStub.pm $
#
package t::FileStub;
use base qw(Bio::Das::Lite);
use strict;
use warnings;
use English qw(-no_match_vars);

our $VERSION = do { my @r = (q$Revision: 19 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub new {
  my ($class, $ref) = @_;
  $ref ||= {};

  my $self = $class->SUPER::new($ref);

  $self->{'filedata'} = $ref->{'filedata'};
  return $self;
}

sub _fetch {
  my ($self, $url_ref, $headers) = @_;

  open my $fh, q(<), $self->{'filedata'} or die "Cannot open $self->{'filedata'}:[$ERRNO]\n";
  local $RS = undef;
  my $xml  = <$fh>;
  close $fh;

  for my $code_ref (values %{$url_ref}) {
    &{$code_ref}($xml);
  }
  return;
}

1;


t/FileStubCompat.pm
#########
# Author:        rmp
# Last Modified: $Date: 2010-03-24 19:29:46 +0000 (Wed, 24 Mar 2010) $ $Author: zerojinx $
# Id:            $Id: FileStubCompat.pm 19 2010-03-24 19:29:46Z zerojinx $
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/t/FileStubCompat.pm,v $
# $HeadURL: https://bio-das-lite.svn.sourceforge.net/svnroot/bio-das-lite/tags/spec-1.53/t/FileStubCompat.pm $
#
package t::FileStubCompat;
use base qw(Bio::DasLite);
use strict;
use warnings;
use English qw(-no_match_vars);

our $VERSION = do { my @r = (q$Revision: 19 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub new {
  my ($class, $ref) = @_;
  $ref ||= {};

  my $self = $class->SUPER::new($ref);

  $self->{'filedata'} = $ref->{'filedata'};
  return $self;
}

sub _fetch {
  my ($self, $url_ref, $headers) = @_;

  open my $fh, q(<), $self->{'filedata'} or die "Cannot open $self->{'filedata'}:[$ERRNO]\n";
  local $RS = undef;
  my $xml  = <$fh>;
  close $fh;

  for my $code_ref (values %{$url_ref}) {
    &{$code_ref}($xml);
  }
  return;
}

1;


t/FileStubChunked.pm
#########
# Author:        rmp
# Last Modified: $Date: 2010-03-24 19:29:46 +0000 (Wed, 24 Mar 2010) $ $Author: zerojinx $
# Id:            $Id: FileStubChunked.pm 19 2010-03-24 19:29:46Z zerojinx $
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/t/FileStubChunked.pm,v $
# $HeadURL: https://bio-das-lite.svn.sourceforge.net/svnroot/bio-das-lite/tags/spec-1.53/t/FileStubChunked.pm $
#
package t::FileStubChunked;
use strict;
use warnings;
use base qw(t::FileStub);

our $VERSION = do { my @r = (q$Revision: 19 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub _fetch {
  my ($self, $url_ref, $headers) = @_;

  open my $fh, q(<), $self->{'filedata'};
  while(my $xml = <$fh>) {
    for my $code_ref (values %{$url_ref}) {
      &{$code_ref}($xml);
    }
  }
  close $fh;
  return;
}

1;


Makefile.PL
#########
# Author:        rmp
# Last Modified: $Date: 2009/01/12 10:29:26 $ $Author: rmp $
# Id:            $Id: Makefile.PL,v 1.47 2009/01/12 10:29:26 rmp Exp $
# Source:        $Source: /cvsroot/Bio-DasLite/Bio-DasLite/Makefile.PL,v $
# $HeadURL$
#
package Makefile;
use strict;
use warnings;
use 5.005;
use ExtUtils::MakeMaker;
use English qw(-no_match_vars);

our $VERSION = do { my @r = (q$Revision: 1.47 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

WriteMakefile(
	      'NAME'          => 'Bio::Das::Lite',
	      'AUTHOR'        => q(Roger Pettett <rpettett@cpan.org>),
	      'ABSTRACT_FROM' => 'lib/Bio/Das/Lite.pm',
	      'VERSION_FROM'  => 'lib/Bio/Das/Lite.pm',
	      'DISTNAME'      => 'Bio-Das-Lite',
	      'PREREQ_PM'     => {
				  'WWW::Curl::Simple' => '0.100181',
				  'HTTP::Request'     => '1.4',
				  'HTTP::Headers'     => '1.64',
				  'SOAP::Lite'        => '0.69',
				  'English'           => '1.01',
				  'Carp'              => '1.02',
				  'Readonly'          => '1.03',
				 },
	      'dist'          => {
				  COMPRESS => 'gzip',
				  SUFFIX   => 'gz',
				 },
	      'LICENSE'       => 'perl',
	     );

print q(
******************************************************************************
** Some of the tests for this module need internet access                   **
** Also remember to set your http_proxy environment variable if you need to **
******************************************************************************
);

1;


lib/Bio/DasLite.pm
#########
# Author: rmp
# Last Modified: $Date $
# Id:            $Id: DasLite.pm,v 1.46 2007/03/16 13:34:39 rmp Exp $
# $HeadURL$
#
# Note: This namespace is now deprecated. Please use Bio::Das::Lite instead.
#
package Bio::DasLite;
use strict;
use warnings;
use base qw(Bio::Das::Lite);

our $VERSION  = do { my @r = (q$Revision: 1.46 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };
*Bio::DasLite::DEBUG = *Bio::Das::Lite::DEBUG;

1;

__END__
=head1 NAME

Bio::DasLite - a compatibility wrapper around Bio::Das::Lite

=head1 VERSION

$Revision: 1.46 $

=head1 DESCRIPTION

  The Bio::DasLite namespace is deprecated. Bio::DasLite is a wrapper for Bio::Das::Lite

=head1 SYNOPSIS

See Bio::Das::Lite

=head1 SUBROUTINES/METHODS

See Bio::Das::Lite

=head1 DIAGNOSTICS

See Bio::Das::Lite

=head1 CONFIGURATION AND ENVIRONMENT

See Bio::Das::Lite

=head1 DEPENDENCIES

See Bio::Das::Lite

=head1 INCOMPATIBILITIES

See Bio::Das::Lite

=head1 BUGS AND LIMITATIONS

See Bio::Das::Lite

=head1 AUTHOR

See Bio::Das::Lite

=head1 LICENSE AND COPYRIGHT

See Bio::Das::Lite

=cut


lib/Bio/Das/Lite/UserAgent.pm
#########
# Author:        rpettett@cpan.org
# Maintainer:    rpettett@cpan.org
# Created:       2005-08-23
# Last Modified: $Date: 2008/12/01 16:59:40 $ $Author: rmp $
# Source:        $Source $
# Id:            $Id $
# $HeadURL $
#
package Bio::Das::Lite::UserAgent;
use strict;
use warnings;
use base qw(LWP::Parallel::UserAgent);
use Readonly;
our $VERSION  = do { my @r = (q$Revision: 1.8 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

Readonly::Scalar our $STATUS_TEXT => {
				      200 => '200 OK',
				      400 => '400 Bad command (command not recognized)',
				      401 => '401 Bad data source (data source unknown)',
				      402 => '402 Bad command arguments (arguments invalid)',
				      403 => '403 Bad reference object',
				      404 => '404 Requested object unknown',
				      405 => '405 Coordinate error',
				      500 => '500 Server error',
				      501 => '501 Unimplemented feature',
				     };

sub new {
  my ($class, %args) = @_;
  my $self = LWP::Parallel::UserAgent->new(%args);
  bless $self, $class;
  return $self;
}

sub on_failure {
  my ($self, $request, $response, $entry) = @_;
  $self->{statuscodes}                  ||= {};

  if (my $das_status = $response->header('X-DAS-Status')) {
    $self->{statuscodes}->{$request->url()} = $STATUS_TEXT->{ $das_status } || $das_status;

  } else {
    $self->{statuscodes}->{$request->url()} = $response->status_line();
  }

  return 1;
}

sub on_return {
  my @args = @_;
  return on_failure(@args);
}

sub statuscodes {
  my ($self, $url)       = @_;
  $self->{statuscodes} ||= {};

  return $url?$self->{statuscodes}->{$url}:$self->{statuscodes};
}

1;

__END__

=head1 NAME

Bio::Das::Lite::UserAgent - A derivative of LWP::Parallel::UserAgent for Bio::Das::Lite use

=head1 VERSION

$Revision: 1.8 $

=head1 SYNOPSIS

=head1 DESCRIPTION

A subclass of LWP::Parallel::UserAgent supporting forward proxies

=head1 SUBROUTINES/METHODS

=head2 new : Constructor

Call with whatever LWP::P::UA usually has

=head2 on_failure : internal error propagation method

=head2 on_return : internal error propagation method

=head2 statuscodes : helper for tracking response statuses keyed on url

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

=head1 DEPENDENCIES

LWP::Parallel::UserAgent

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

=head1 AUTHOR

Roger Pettett, E<lt>rpettett@cpan.orgE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2007 GRL, by Roger Pettett

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/Bio/Das/Lite/UserAgent/proxy.pm
#########
# Author:        rmp@sanger.ac.uk
# Maintainer:    rmp@sanger.ac.uk
# Created:       2005-08-23
# Last Modified: $Date: 2007/10/22 08:35:07 $ $Author: rmp $
# Source:        $Source $
# Id:            $Id $
# $HeadURL $
#
package Bio::Das::Lite::UserAgent::proxy;
use strict;
use warnings;

our $VERSION  = do { my @r = (q$Revision: 1.2 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub host     { my $self = shift; return $self->{'host'}; }
sub port     { my $self = shift; return $self->{'port'}; }
sub scheme   { my $self = shift; return $self->{'scheme'}; }

#########
# userinfo, presumably for authenticating to the proxy server.
# Not sure what format this is supposed to be (username:password@ ?)
# Things fail silently if this isn't present.
#
sub userinfo { return q(); }

1;
__END__

=head1 NAME

Bio::Das::Lite::UserAgent::proxy - A derivative of LWP::Parallel::UserAgent for Bio::Das::Lite use

=head1 VERSION

$Revision: 1.2 $

=head1 SYNOPSIS

=head1 DESCRIPTION

A support class for information about a forward proxy server

=head1 SUBROUTINES/METHODS

=head2 host : get/set host

=head2 port : get/set port

=head2 scheme : get/set scheme

=head2 userinfo : stub for authentication? Stops LWP::P::UA from silently failing

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

=head1 DEPENDENCIES

LWP::Parallel::UserAgent

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

=head1 AUTHOR

Roger Pettett, E<lt>rmp@sanger.ac.ukE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2007 GRL, by Roger Pettett

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

=cut


Build.PL
#########
# Author:        rmp
# Last Modified: $Date: 2010-04-16 12:29:45 +0100 (Fri, 16 Apr 2010) $
# Id:            $Id: Build.PL 32 2010-04-16 11:29:45Z andyjenkinson $
# Source:        $Source$
# $HeadURL: https://bio-das-lite.svn.sourceforge.net/svnroot/bio-das-lite/tags/spec-1.53/Build.PL $
#
use strict;
use warnings;
use Module::Build;
use lib qw(lib);

my $builder = Module::Build->new(
				 module_name         => 'Bio::Das::Lite',
				 dist_author         => 'Roger Pettett <rmp@psyphi.net>',
				 dist_abstract       => q[Bio::Das::Lite is an implementation of the BioDas protocol, http://biodas.org/],
				 license             => 'perl',
				 requires            => {
							 'WWW::Curl'         => '4.07',
							 'HTTP::Response'    => '1.49',
							 'English'           => '1.01',
							 'Carp'              => '1.02',
							 'Readonly'          => '1.03',
							},
				 recommends          => {
							},
				 build_requires      => {
							 'Test::More'              => '0.8',
							},
				 build_recommends    => {
							 'Test::Distribution'      => '2.00',
							 'Test::Perl::Critic'      => '1.01',
							},
				 sign                => 1,
				);

$builder->create_build_script();

print q(
******************************************************************************
** Some of the tests for this module need internet access                   **
** Also remember to set your http_proxy environment variable if you need to **
******************************************************************************
);
